File: scripts/repo_size.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import subprocess
import sys
from pathlib import Path


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Measure repository file size using git-tracked and unignored files only."
        )
    )
    parser.add_argument(
        "--exclude",
        action="append",
        default=["demo"],
        help="Top-level path prefix to exclude. Can be passed multiple times.",
    )
    return parser.parse_args()


def format_size(size_bytes: int) -> str:
    mib = size_bytes / 1024 / 1024
    return f"{size_bytes} bytes ({mib:.2f} MiB)"


def main() -> int:
    args = parse_args()
    root = Path.cwd()
    exclude_prefixes = tuple(
        prefix.strip("/\\") for prefix in args.exclude if prefix.strip("/\\")
    )

    try:
        result = subprocess.run(
            ["git", "ls-files", "-z", "--cached", "--others", "--exclude-standard"],
            capture_output=True,
            text=False,
            check=True,
        )
    except FileNotFoundError:
        print("git was not found on PATH.", file=sys.stderr)
        return 1
    except subprocess.CalledProcessError as exc:
        sys.stderr.write(exc.stderr.decode("utf-8", "replace"))
        return exc.returncode

    total_size = 0
    file_count = 0

    for raw_path in result.stdout.split(b"\0"):
        if not raw_path:
            continue

        rel_path = Path(raw_path.decode("utf-8", "surrogateescape"))
        if rel_path.parts and rel_path.parts[0] in exclude_prefixes:
            continue

        abs_path = root / rel_path
        if not abs_path.is_file():
            continue

        total_size += abs_path.stat().st_size
        file_count += 1

    print(f"Files counted: {file_count}")
    print(f"Excluded prefixes: {', '.join(exclude_prefixes) if exclude_prefixes else '(none)'}")
    print(f"Total size: {format_size(total_size)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/repo-size.mjs
import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import { execFileSync } from "node:child_process";

function parseArgs(argv) {
  const excludes = ["demo"];
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--exclude") {
      const value = argv[index + 1];
      if (!value) {
        throw new Error("Missing value after --exclude");
      }
      excludes.push(value);
      index += 1;
    }
  }
  return {
    excludes: excludes
      .map((value) => value.replace(/[\\/]+$/u, "").replace(/^[\\/]+/u, ""))
      .filter(Boolean)
  };
}

function formatBytes(bytes) {
  return `${bytes} bytes (${(bytes / 1024 / 1024).toFixed(2)} MiB)`;
}

function isExcluded(relativePath, excludes) {
  const normalized = relativePath.split(path.sep).join("/");
  return excludes.some((prefix) => normalized === prefix || normalized.startsWith(`${prefix}/`));
}

function listGitFiles() {
  const output = execFileSync(
    "git",
    ["ls-files", "-z", "--cached", "--others", "--exclude-standard"],
    { encoding: "utf8" }
  );
  return output.split("\0").filter(Boolean);
}

function main() {
  const { excludes } = parseArgs(process.argv.slice(2));
  const files = listGitFiles();

  let count = 0;
  let totalSize = 0;

  for (const relativePath of files) {
    if (isExcluded(relativePath, excludes)) {
      continue;
    }

    let stat;
    try {
      stat = fs.statSync(relativePath);
    } catch {
      continue;
    }

    if (!stat.isFile()) {
      continue;
    }

    count += 1;
    totalSize += stat.size;
  }

  console.log(`Files counted: ${count}`);
  console.log(`Excluded prefixes: ${excludes.length > 0 ? excludes.join(", ") : "(none)"}`);
  console.log(`Total size: ${formatBytes(totalSize)}`);
}

try {
  main();
} catch (error) {
  const message = error instanceof Error ? error.message : String(error);
  console.error(message);
  process.exit(1);
}
